blocks/model.py

```diff
@@ -0,0 +1,181 @@
+        ----------
+        param_values : OrderedDict
+            Dictionary of (parameter name, :class:`~numpy.ndarray`) pairs.
+
+        """
+        params = self.get_params()
+
+        unknown = set(param_values) - set(params)
+        missing = set(params) - set(param_values)
+        if len(unknown):
+            logger.error("unknown parameter names:\n".format(unknown))
+        if len(missing):
+            logger.error("missing values for parameters:\n".format(missing))
+
+        for name, value in param_values.items():
+            if name in params:
+                params[name].set_value(value)
+
+    @abstractmethod
+    def get_criterion(self):
+        """Return the optimization criterion."""
+        pass
+
+    @abstractmethod
+    def get_top_bricks(self):
+        """Return the top-level bricks that are used in the model.
+
+        Returns
+        -------
+        bricks : list
+            List of bricks.
+
+        """
+        pass
+
+
+class Model(AbstractModel, ComputationGraph):
+    """Wraps a computation graph to support model interface.
+
+    This model covers the most common case when all information
+    about the model is contained in an annotated computation graph:
+    parameters are identified by the roles, bricks found by annotations.
+    Due to frequency of this case this class is called simply 'Model'
+    and not 'ComputationGraphModel'.
+
+    Parameters
+    ----------
+    outputs : (list of) :class:`~theano.Variable`
+        The output variables of the computation graph.
+
+    .. todo::
+
+        Overriding the automatically found parameters and bricks might
+        be needed.
+
+        If there top bricks in scan inner graphs, those will not be found.
+
+    """
+    def __init__(self, outputs):
+        super(Model, self).__init__(outputs)
+
+        bricks = [get_brick(var) for var in self if get_brick(var)]
+        children = set(chain(*(brick.children for brick in bricks)))
+        # Quadratic complexity: we should not have thousands of
+        # top-level bricks.
+        self.top_bricks = []
+        for brick in bricks:
+            if brick not in children and brick not in self.top_bricks:
+                self.top_bricks.append(brick)
+
+        brick_param_names = {
+            v: k for k, v in Selector(self.top_bricks).get_params().items()}
+        self.params = []
+        for param in VariableFilter(roles=[PARAMETER])(self.shared_variables):
+            if param in brick_param_names:
+                self.params.append((brick_param_names[param], param))
+            else:
+                self.params.append((param.name, param))
+        self.params = OrderedDict(self.params)
+
+    def get_criterion(self):
+        """Return the output variable, if there is a single one.
+
+        If there is only one output variable, it is a reasonable default
+        setting to assume that it is the optimization objective.
+
+        """
+        if len(self.outputs) == 1:
```

Comment:
All I'm saying is that in that 1% case where someone passes multiple outputs to `__init__`, we know at that point in time that calls to `get_criterion` are going to fail in the future. Yet, we allow `Model` to be initialized, potentially 100 epochs of training being run, before some extension calls `get_criterion` and crashes. All I'm saying is that it might be safer to just raise an error in `__init__` then, or to support passing multiple outputs fully.